lib/astc_encode.cpp

```diff
@@ -0,0 +1,963 @@
+
+        ktx_size_t diff = level_offset - prev_level_offset;
+
+        assert(diff - priv._levelIndex[level].byteLength == 0
+               && "\nSome format/blocks aren't alighned to 8 bytes, need to create offsets\n");
+    }
+
+    free(This->pData); // No longer needed. Reduce memory footprint.
+    This->pData = NULL;
+    This->dataSize = 0;
+
+    // Allocate big enough buffer for all compressed images
+    This->dataSize = astc_images_size;
+    This->pData = reinterpret_cast<uint8_t *>(malloc(This->dataSize));
+    if (!This->pData) {
+        return KTX_OUT_OF_MEMORY;
+    }
+
+    uint8_t* buffer_out  = This->pData;
+    uint32_t input_size = astc_input_images.size();
+
+    for (uint32_t level = 0; level < input_size; level++) {
+        if (params->verbose)
+            std::cout << "Astc compressor: compressing image = " <<
+                         level + 1 << " of " << input_size  << std::endl;
+
+        // Lets compress to astc
+        astcenc_image *input_image = astc_input_images[level];
+        ktx_size_t     size        = astc_image_sizes[level];
+
+        CompressionWorkload work;
+        work.context = astc_context;
+        work.image = input_image;
+        work.swizzle = swizzle;
+        work.data_out = buffer_out;
+        work.data_len = size;
+        work.error = ASTCENC_SUCCESS;
+
+        launchThreads(threadCount, compressionWorkloadRunner, &work);
+
+        if (work.error != ASTCENC_SUCCESS) {
+            std::cout << "Astc compressor failed\n" <<
+                         astcenc_get_error_string(work.error) << std::endl;
+
+            for(auto& ii : astc_input_images)
+                imageFree(ii);
+
+            astcenc_context_free(astc_context);
+            return KTX_INVALID_OPERATION;
+        }
+
+        buffer_out += size;
+
+        // Free input image
+        imageFree(input_image);
+
+        // Reset astc context for next image
+        astcenc_compress_reset(astc_context);
+    }
+
+    // We are done with astcencoder
+    astcenc_context_free(astc_context);
+
+    ktx_uint32_t transfer = KHR_DFDVAL(BDB, TRANSFER);
+    bool sRGB = transfer == KHR_DF_TRANSFER_SRGB; // Is this right or I need "KHR_DF_PRIMARIES_SRGB"?
+
+    This->vkFormat = astcVkFormat(params->blockSize, sRGB);
+
+    free(This->pDfd);
+    This->pDfd = vk2dfd(static_cast<VkFormat>(This->vkFormat));
+
+    // reset pointer to new BDB
+    BDB = This->pDfd+1;
+
+    // Astc-related checks
+    ktx_uint32_t model = KHR_DFDVAL(BDB, MODEL);
+    ktx_uint32_t primaries = KHR_DFDVAL(BDB, PRIMARIES);
+    This->supercompressionScheme = KTX_SS_ASTC;
+    This->isCompressed = true;
+
+    assert(model == KHR_DF_MODEL_ASTC && "Invalid dfd generated for astc image\n");
+    if (transfer == KHR_DF_TRANSFER_SRGB) {
+        assert(primaries == KHR_DF_PRIMARIES_SRGB && "Not a valid sRGB image\n");
+    }
+
+    // Since we only allow 8-bit components to be compressed
```

Comment:
The comment is wrong. It should be "Block-compressed textures never need byte swapping so typeSize is 1."

If you copied this from ktxTexture2_CompressBasisEx, please fix it there too.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/c01d05d4cfa0353f848c18051d90ba94862115b5

Reply:
Later its removed after prototype texture implementation.